_build/man/animated_graph.1
.\" Man page generated from reStructuredText.
.
.TH "ANIMATED_GRAPH" "1" "May 13, 2015" "0.0.1" "Animated_graph"
.SH NAME
animated_graph \- Animated_graph Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH PROGRAM DEPENDENCIES:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
python 3
.IP \(bu 2
gnuplot
.IP \(bu 2
ffmpeg
.UNINDENT
.UNINDENT
.UNINDENT
.SH CONTENTS:
.SS Animated_graph Tutorial
.sp
This is short desription about how to use animeted_graph
.INDENT 0.0
.TP
.B \fBusage\fP: animated_graph [\-h] [\-f FILE] [\-t TIMEFORMAT] [\-x XMIN] [\-X XMAX]
.INDENT 7.0
.INDENT 3.5
[\-y YMIN] [\-Y YMAX] [\-S SPEED] [\-T TIME] [\-F FPS]
[\-l LEGEND] [\-g GNUPLOTPARAMS [GNUPLOTPARAMS ...]]
[\-e EFFECTPARAMS] [\-n NAME]
file [file ...]
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.BI \-f \ FILE\fP,\fB \ \-\-conf_file \ FILE
Specify config file.
.TP
.BI \-t \ TIMEFORMAT\fP,\fB \ \-\-TimeFormat
Specify timeformat which is udes in data file (default:"[%Y/%m/%d %H:%M:%S]").
.TP
.BI \-x \ XMIN\fP,\fB \ \-\-Xmin \ XMIN
Specify  minimal x value, can be int/float value or keyword:min (default:min).
.TP
.BI \-X \ XMAX\fP,\fB \ \-\-Xmax \ XMAX
Specify  maximal x value, can be int/float value or keyword:max (default:max).
.TP
.BI \-y \ YMIN\fP,\fB \ \-\-Ymin \ YMIN
Specify  minimal y value, can be int/float value or keyword:min (default:min).
.TP
.BI \-Y \ YMAX\fP,\fB \ \-\-Ymax \ YMAX
Specify  maximal y value, can be int/float value or keyword:max (default:max).
.TP
.BI \-S \ SPEED\fP,\fB \ \-\-Speed \ SPEED
Specify speed of animation (default:1).
.TP
.BI \-T \ TIME\fP,\fB \ \-\-Time \ TIME
Specify duration of animations in seconds.
.TP
.BI \-F \ FPS\fP,\fB \ \-\-FPS \ FPS
Specify FPS (frames per second) of animations (default:25).
.TP
.BI \-l \ LEGEND
Specify legend (title) if animation
.TP
.BI \-g \ GNUPLOTPARAMS
Specify parametrs which will be passed to gnuplot (only set parameters) any previous setting will be overdriven .
.TP
.BI \-e \ EFFECTPARAMS
Specify parameters of effect which are : ColorMadness=true/false,ColorUp=color,ColorDown=color,ColorMiddle=color,
color=[red,blue etc..], (parameters are separated with \(aq:\(aq)
.TP
.BI \-n \ NAME\fP,\fB \ \-\-Name \ NAME
Specify name of file to be produced wtih extension (name.mp4)
.TP
.B \-v\fP,\fB  \-\-verbouse
verbouse (default: false)
\-h, \-\-help                                                                           show this help message and exit
.UNINDENT
.UNINDENT
.SS Example of usage in command line:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\&./arg_test.py data \-T 10 \-F 25 \-t "[%Y/%m/%d %H:%M:%S]" \-Y max \-y min \-e ColorMadness=true \-X "[2009/05/12 07:29:00]" \-x "[2009/05/11 07:30:00]" \-n simple \-g tics
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.SS Example of config file:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
[Defaults]
#This need to be here

#This is signle line comment
# Graph animation : Simple Effect
# Example config file

Timeformat [%Y/%m/%d %H:%M:%S]

Xmin [2009/05/11 07:30:00]

Xmax [2009/05/12 07:29:00]

Ymin \-1

Ymax 1

Speed 20

FPS 25

Legend Test animation

Gnuplotparams title grid

Effectparams ColorUp=red:ColorDown=green

Name test
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBExample of data file:\fP
.INDENT 0.0
.INDENT 3.5
Data are separate with spaces , last collum is considered as Y\-value rest is dataformat as specified in \fITimeFormat\fP
.UNINDENT
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
[2009/05/11 07:30:00] 0
[2009/05/11 07:31:00] 0.00999983333416666468
[2009/05/11 07:32:00] 0.01999866669333307936
[2009/05/11 07:33:00] 0.02999550020249566076
[2009/05/11 07:34:00] 0.03998933418663415945
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Documentation
.INDENT 0.0
.TP
.B animated_graph.check_fts(speed, fps, time)
Check exclusivity of sped,fps,time.
.INDENT 7.0
.TP
.B \fBArgs:\fP
.INDENT 7.0
.INDENT 3.5
\fIspeed (float)\fP: speed to check
.sp
\fIfps (float)\fP: fps to check
.sp
\fItime (float)\fP: time to check
.UNINDENT
.UNINDENT
.sp
Program halt when there is only one value passed from user.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B animated_graph.check_time(data, format)
.INDENT 7.0
.TP
.B \fBArgs:\fP
\fIdata (str)\fP: string to check_time
.sp
\fIformat (str)\fP: format to check
.TP
.B \fBReturns:\fP
\fItime.strptime\fP\&. The return code of this function is time.strptime.
If string cannot by coverted to time.strptime program halt.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B animated_graph.check_value(tocheck)
.INDENT 7.0
.TP
.B \fBArgs:\fP
\fItocheck (str)\fP: string to check_value
.TP
.B \fBReturns:\fP
\fIfloat\fP\&. The return code of this function is string coverted to float value.
If string cannot by coverted to float program halt.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B animated_graph.effect(effectpar)
.INDENT 7.0
.INDENT 3.5
This function process data from effect parameters
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B \fBArgs:\fP
\fIeffectpar (str)\fP: string to parse
.UNINDENT
.sp
\fBReturns:\fP
.INDENT 7.0
.INDENT 3.5
\fIstring ColorMadness\fP \- if effect ColorMadness should be applied
.sp
\fIstring ColorUp\fP     \-color of inverse bars above zero
.sp
\fIstring ColorDown\fP   \-color of inverse bars belowe zero
.sp
\fIstring ColorMiddle\fP \-color of actuall data
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B animated_graph.main()
Generate animation from given data; return a video file
.UNINDENT
.INDENT 0.0
.TP
.B animated_graph.merge_files(file_data)
Merge input files to one temporary file
.INDENT 7.0
.TP
.B \fBArgs:\fP
\fIfile_data (str)\fP: data from file
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B animated_graph.parse_effects(string)
.INDENT 7.0
.INDENT 3.5
This is custom type for argparser
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B \fBArgs:\fP
\fIstring (str)\fP: string to parse
.TP
.B \fBReturns:\fP
\fIstring\fP\&. The return code of this function is string
If string cannot by coverted to time.strptime program halt.
.TP
.B \fBRaises:\fP
\fIargparse.ArgumentTypeError\fP
.UNINDENT
.UNINDENT
.SS Developer documentation
.sp
This document descibes all program opertions and tries to explain logic behind them.
.SS Overview
.sp
Program firstly parse user input from line or from config file. In case when there are both, we save parameters from config file as defaults and overide them with paramaters from command line. For parsing we use standart libraries. \fIArgparser\fP handle parsing paramaters from command line and \fIConfigparser\fP parsing config files.
.sp
Functions \fIparse_effects\fP and \fIeffect\fP are used for additional parsing of effect paramaters in format described in manual.
.sp
After program identifies user input paramters we procced to read data file from file or url.
All temp files will be located in directory created by \fItempfile.mkdtemp()\fP .
.sp
In case of multiple input files, program merge them to one temp file using function \fImerge_files(file_data)\fP\&. File data is list containing input files.
This function is called even if there is only one input file.
.sp
Function then sort values in time using X data.
.sp
After that function tries to find duplicties. Duplicty is considered either X value with multiple Y values or multiple X values with same Y value.Function keeps only first occurences of X value and ignores others. After that function writes this values to temp file.
.sp
From now on program use only generated temp file for all operations.
.sp
We check file for bad Y values using function \fIcheck_value(tocheck)\fP
which tries to convert string to float, if conversion fails function print error messaage and quit.Program assume that Y data are only in last collum (in line separated with spaces), processing more then one Y value per line is not supported.
.sp
Similarly using functuion \fIcheck_time(data,format)\fP program tries to convert string to time_struct using \fItime.strptime(string,format)\fP\&.
In this case \fIformat\fP is either default time format or format provided by user.
.sp
If bad value ( wrong time format, wrong value ) is found , proram prints error message clean tempfiles and quits.
.sp
Now we need to calculate how many frames we have to plot.Program use default values if user didnt specified otherwise.
.sp
For our animation effect we draw all X values at once with diffrent Y values for each frame. For each frame we generate new configuration file for \fIGnuplot\fP\&. If user defined any paramaters for gnuplot we add them to configuration file.
For actuall plotting there is hardcoced some logic for gnuplot, generally we differentiate 3 diffrent scenarios.
Ploting only on positive Y axes , only on negative Y axes or on both at the same time. We need to know that for our inversion effect.
.sp
After we generated all frames with gnuplot, program calls \fIffmpeg\fP to create video in mp4 format. Program cleans all temp files and quits.
.SS Data Inputs
.sp
Data inputs are single/multiple files or urls. They need to have two collums, first collum is considered as X value for ploting, in this case Time.In second collum they can be any number values.If duplicity is found, only first occurence is saved, rest are ignored.
.SS Outputs
.sp
Output for this program is one video file in user defined location. If program ends with error message, no output file is generated. If file with output name already exist, program adds \fI_i\fP at the end. Where \fIi\fP is increasing sequence of integers.
.SS Arguments
.sp
Program accepts several arguments from command line, detaily described in \fItutorial\fP, if there is more then one occurance of given argument, only last is used, rest of them are ignored. If some of the optional arguments are not defined by user, program use default values.
.SS Configuration file
.sp
Program can proccess configuration file from user defined location. Arguments from command line have priority over arguments defined in configuration file.Program does not support more than one configuration file at the time.
.INDENT 0.0
.IP \(bu 2
\fIgenindex\fP
.IP \(bu 2
\fImodindex\fP
.IP \(bu 2
\fIsearch\fP
.UNINDENT
.SH AUTHOR
Lubos Hercka
.SH COPYRIGHT
2014, Lubos Hercka
.\" Generated by docutils manpage writer.
.


script/animated_graph.py
#! /usr/bin/env python3
from subprocess import call
import argparse
import sys
import time
import signal
import configparser
import tempfile
import shutil
import os
from os import listdir
import subprocess
import random
import urllib.request
import unittest


def check_value(tocheck):
  """
  **Args:**
    *tocheck (str)*: string to check_value

  **Returns:**
    *float*. The return code of this function is string coverted to float value.
    If string cannot by coverted to float program halt.
  """
  try:
    return float (tocheck)
  except ValueError:
    print ("Error in data file")
    shutil.rmtree(dirpath) 
    sys.exit(1)

def check_fts (speed,fps,time):
  """ Check exclusivity of sped,fps,time.

  **Args:**
    *speed (float)*: speed to check

    *fps (float)*: fps to check

    *time (float)*: time to check

   Program halt when there is only one value passed from user.
  """
  count=0
  if not speed:
    count+=1
  if not fps:
    count+=1
  if not time:
    count+=1
  if count==2:
    print ("Invalid fps|speed|time !")
    shutil.rmtree(dirpath)
    sys.exit(1)
def check_time (data,format):
  """
  **Args:**
    *data (str)*: string to check_time

    *format (str)*: format to check

  **Returns:**
    *time.strptime*. The return code of this function is time.strptime.
    If string cannot by coverted to time.strptime program halt.
  """
  try:
    return time.strptime(data,format)
  except ValueError:
    print ("Bad time format")
    shutil.rmtree(dirpath)
    sys.exit(1)
#def effect_params (string):
def parse_effects (string):
  """
   This is custom type for argparser
  **Args:**
    *string (str)*: string to parse

  **Returns:**
    *string*. The return code of this function is string
    If string cannot by coverted to time.strptime program halt.

  **Raises:**
       *argparse.ArgumentTypeError*
  """
     
  effect_par=['ColorMadness','ColorUp','ColorDown','ColorMiddle']
  ef_par=''.join(string)
  ef_par=ef_par.split(':')
  a=all( y.split('=')[0] in effect_par for y in ef_par)     
  b=all(len(y.split('='))==2 for y in ef_par)
  if not a or not b:
    raise argparse.ArgumentTypeError("Invalid Effect parameters")
  else:
   return string
def merge_files(file_data):
  """
  Merge input files to one temporary file

  **Args:**
    *file_data (str)*: data from file

  """
  #data = file_data.readlines()
  
  data = []
  with open(dirpath+'/tmp_data', 'w') as outfile:
    for x in file_data:
      data += x.readlines()
        
    data.sort(key = lambda x: x.split()[:-1])

     #Resolve duplicates
    out = []
    for x in data:
      if x.split()[:-1] not in [y.split()[:-1] for y in out]:
        out.append(x)

    for line in out:
      outfile.write(line)


    #print (data)
    #data.sort(key = lambda x: x.split()[:-1])
   # print (data)

def effect(effectpar):
  """
   This function process data from effect parameters
  **Args:**
    *effectpar (str)*: string to parse

  **Returns:**

    *string ColorMadness* - if effect ColorMadness should be applied

    *string ColorUp*     -color of inverse bars above zero

    *string ColorDown*   -color of inverse bars belowe zero

    *string ColorMiddle* -color of actuall data

  """
  ColorMadness="false"
  ColorUp=""
  ColorDown=""
  ColorMiddle=""

  ef_par=''.join(effectpar)
  ef_par=ef_par.split(':')
  for y in ef_par:
    if y.split('=')[0] =='ColorMadness':
      if y.split('=')[1] == "true" or y.split('=')[1] == "false":
        ColorMadness=y.split('=')[1]
        
    if y.split('=')[0] =='ColorUp':
          ColorUp=y.split('=')[1]

    if y.split('=')[0] =='ColorDown':
         ColorDown=y.split('=')[1]
      
    if y.split('=')[0] =='ColorMiddle':
          ColorMiddle=y.split('=')[1]
  return (ColorMadness,ColorUp,ColorDown,ColorMiddle)  

def url_parser(url):
  if url[0:4]=="http":
    try:
      urllib.request.urlretrieve(url,dirpath+'/tmp_netfile')
    except EnvironmentError:
       print (url,'not found !')
       shutil.rmtree(dirpath)
       sys.exit(1)

    infile = open(dirpath+'/tmp_netfile','r')
    return (infile)
  else:
      try:
        infile = open(url,'r')
        return (infile)
      except EnvironmentError:
        print ('File',url,'not found !')
        shutil.rmtree(dirpath)
        sys.exit(1)

def main():
  '''Generate animation from given data; return a video file''' 
  global dirpath 
  dirpath= tempfile.mkdtemp()
  #Parsing arguments
  conf_parser = argparse.ArgumentParser(
  
    add_help=False
      )
  conf_parser.add_argument("-f", "--conf_file",
  help="Specify config file", metavar="FILE")
  args, remaining_argv = conf_parser.parse_known_args()
  defaults = {
  "xmax" : "max",
  "xmin" : "min",
  "ymax" : "max",
  "ymin" : "min",
  "timeformat" : "[%Y/%m/%d %H:%M:%S]",
  "name" : "skj",
  }


  items=["blue","red","blue","green","orange","purple","yellow"]
  if args.conf_file:
      config = configparser.RawConfigParser(delimiters=(' '), comment_prefixes=('#'))
      config.read([args.conf_file])
      defaults = dict(config.items("Defaults"))
      if "effectparams" in defaults:
        ColorMadness,ColorUp,ColorDown,ColorMiddle=effect(defaults["effectparams"])
        if ColorUp:
          items[0]=ColorUp
        if ColorMiddle:
          items[1]=ColorMiddle
        if ColorDown:
          items[2]=ColorDown


   
  parser = argparse.ArgumentParser(
  parents=[conf_parser],
  description=__doc__,
  formatter_class=argparse.RawDescriptionHelpFormatter,
  )
  parser.set_defaults(**defaults)

  parser.add_argument('data', metavar='file',type=url_parser,help='data to proccess',nargs='+')


  parser.add_argument('-t','--TimeFormat', action='store',dest="timeformat",
                     help='Set value of timestamp format')

  parser.add_argument('-x','--Xmin', action='store',
                     dest='xmin',
                     help='x-min (default: min)')

  parser.add_argument('-X','--Xmax', action='store',
                     dest='xmax',
                     help='x-max (default:  max )')

  parser.add_argument('-y','--Ymin', action='store',
                     dest='ymin',                   
                     help='y-min (default: min)')

  parser.add_argument('-Y','--Ymax', action='store',
                     dest='ymax',                   
                     help='y-max(default:  max)')


  parser.add_argument('-S','--Speed', action='store',type=float,
                     dest='speed',
                     help='speed (default: 1 record/frame)')

  parser.add_argument('-T','--Time', action='store',type=float,
                     dest='time',
                     help='time (duration)')

  parser.add_argument('-F','--FPS', action='store' ,type=float,
                     dest='fps',
                     help='fps (default: 25)')

  parser.add_argument('-l', action='store',
                     dest='legend',
                     help='legend (default: find the max)')

  parser.add_argument('-g','--GnuplotParams', action='store',nargs='*',
                     dest='gnuplotparams',
                     help='GnuplotParams (default: n/a)')

  parser.add_argument('-e', action='store',nargs=1,
                      dest='effectparams',type=parse_effects,
                     help='EffectParams (default: n/a)')

  parser.add_argument('-n','--Name', action='store',
                     dest='name',
                     help='Name (default: n/a)')

  parser.add_argument('-v','--verbouse', action='store_true',
                      dest='verbouse',
                     help='verbouse (default: false)')

  parser.add_argument('-E','--IgnoreErrors', action='store_true',
                      dest='ignore',
                     help='Ignore errrors (default: false)')


  x=0
  frames=0
  args = parser.parse_args(remaining_argv)
  if not args.ymax:
    args.ymax="max"
  if not args.ymin:
    args.ymin="min"
  if not args.xmax:
    args.xmax="max"
  if not args.xmin:
    args.xmin="min"
  if not args.timeformat:
    args.timeformat="[%Y/%m/%d %H:%M:%S]"
  if not args.name:
    args.name="skj"

  #If file already exist add suffix name_i
  temp = args.name.split('/')[:-1]
  if not temp:
    dirr ="."
  else:
    dirr ="".join(temp)

  filenum=[]
  for subor in listdir(dirr):
    if args.name.split('/')[-1] == subor.rsplit("_")[0]:
      num=(subor.rsplit("_")[-1])[:-4]
      filenum.append(num)
    if args.name.split('/')[-1] == subor[:-4]:
      filenum.append('0')
  
  #result = int (max(filenum))
  results = [int(i) for i in filenum]
  if results:
    args.name =args.name + "_" + str(max(results) + 1)

  '''
  suf = 1
  if (os.path.isfile(args.name+'.mp4')):
    while (os.path.isfile(args.name + "_" + str(suf) +'.mp4')):
      suf +=1
    args.name = args.name + "_" + str(suf) 
  '''


  #Merge input files 
  merge_files(args.data)
            
  #Read data from merged input files
  with open(dirpath+'/tmp_data', 'r') as infile:
    data = infile.readlines()
  #Number of rows in input files
  rows_n=len(data)
  if rows_n==0:
    print ("Empty file !")
    shutil.rmtree(dirpath)
    sys.exit(1)

  #Checking input data file (correct fromat and type of value)
  times=[]
  value=[]
  
  for row in data:
        if row.strip():
          #print (row)
          tmfrmt=row.split()[:-1]
          collum=len(row.split())
          tmfrmt=" ".join(tmfrmt)
          value.append(check_value (row.split()[-1]))
          times.append(time.strftime(args.timeformat,check_time(tmfrmt,args.timeformat)))


  # Finding Ymax
  if args.ymax == "max" : 
        ymax=float (max(value))
  else:
    ymax=check_value(args.ymax)
       
  # Finding Ymin
  if args.ymin == "min" :
        ymin=float (min(value))
  else:
    ymin=check_value(args.ymin)
  
  # Finding Xmax
  if args.xmax == "max" :
        xmax=max(times)
  else:
    xmax=time.strftime(args.timeformat,check_time(args.xmax,args.timeformat))
  
  # Finding Xmin
  if args.xmin == "min" :
        xmin =min(times)
  else:
    xmin=time.strftime(args.timeformat,check_time(args.xmin,args.timeformat))

  if args.fps and args.speed and args.time:
    frames=rows_n/args.speed
    if args.time != frames/args.fps:
      parser.print_usage()
      print ("Invalid fps|speed|time")
      shutil.rmtree(dirpath)
      sys.exit(1)


  # Calculating how many frames have to be generated .
  #check_fts(args.speed,args.fps,args.time)
  if not args.fps  and args.speed and args.time:
    frames=rows_n/args.speed
    args.fps=frames/args.time

  if args.fps  and not args.speed and args.time:
    frames=args.fps * args.time
    args.speed = rows_n /frames

  if args.fps and args.speed and not args.time:
    frames=rows_n/args.speed
    args.time = frames / args.fps



  if args.fps and not args.speed and not args.time:
    args.speed = 1
    frames=rows_n/args.speed
    args.time = frames / args.fps

  if not args.fps and args.speed and not args.time:
    args.fps = 25
    frames=rows_n/args.speed
    args.time = frames / args.fps

  if not args.fps and not args.speed and args.time:
    args.fps = 25
    frames=args.fps * args.time
    args.speed = rows_n /frames


  if not frames :
    args.fps=25
    args.speed=1
    frames=rows_n/args.speed
    args.time = frames / args.fps
  

  #Printing basic information when --verbouse is true
  if args.verbouse:
    print ("FPS: ","%.2f" %args.fps)
    print ("TIME: ","%.2f" %args.time,"s")
    print ("SPEED: ","%.2f" %args.speed)
    print ("ROWS: ",rows_n)
    print ("FRAMES: ","%.2f" %frames)

  if isinstance(args.gnuplotparams, str):
    args.gnuplotparams=args.gnuplotparams.split()

  #Prepare Gnuploparameters to set in gnuplot
  if args.gnuplotparams:
    toset=['set '+y+'\n' for y in args.gnuplotparams]
    toset= ''.join(toset)
  else:
    toset=''
  if not args.legend:
    args.legend = ""

  if args.effectparams:
    #Assign values from function wich proceese effect parameters
    ColorMadness,ColorUp,ColorDown,ColorMiddle=effect(args.effectparams)  
    #List of colors for bars
    if ColorUp:
      items[0]=ColorUp
    if ColorMiddle:
      items[1]=ColorMiddle
    if ColorDown:
      items[2]=ColorDown

  if args.verbouse:
    print ("Generetaing frames...\n")
  #Main loop;each iteration generate temp gnuplot configuration file and call gnuplot to process data
  for x in range(0, int(frames)):
        rnd=dirpath+'/'+str(x)+'.png'
        filename=dirpath+'/temp.gp'
        if args.effectparams:
          if ColorMadness=="true":
            random.shuffle(items)

        if ymin>=0 and ymax>0:
            plot='plot  "'+str(infile.name)+'" every ::0::'+str(rows_n)+' using 1:( '+str(ymin)+'+( ( ($'+str(collum)+'-'+str(ymin)+')/'+str(frames)+')*' +str(x)+') ) axes x1y1 with boxes notitle linecolor rgb "'+items[1]+'",\
            "'+str(infile.name)+'" every ::0::'+str(rows_n)+\
            'using 1:(($'+str(collum)+'*0)+'+str(ymax)+'):('+str(ymax)+'-(('+str(ymax)+'-(($'+str(collum)+' >= 0)  ? $'+str(collum)+' : $'+str(collum)+'*0))/'+str(frames)+')*'+str(x)+'):(($'+str(collum)+'*0)+'+str(ymax)+'):('+str(ymax)+'-(('+str(ymax)+'-$'+str(collum)+')/'+str(frames)+')*'+str(x)+')\
            axes x1y1 with candlesticks notitle linecolor rgb "'+items[0]+'"\
            '
        if ymin<0 and ymax<=0:
            plot='plot  "'+str(infile.name)+'" every ::0::'+str(rows_n)+' using 1:((($'+str(collum)+'/'+str(frames)+')*' +str(x)+') ) axes x1y1 with boxes notitle linecolor rgb "'+items[1]+'",\
            "'+str(infile.name)+'" every ::0::'+str(rows_n)+\
            'using 1:(($'+str(collum)+'*0)'+str(ymin)+'):('+str(ymin)+'-(('+str(ymin)+'-(($'+str(collum)+' <= 0)  ? $'+str(collum)+' : $'+str(collum)+'*0))/'+str(frames)+')*' +str(x)+'):(($'+str(collum)+'*0)'+str(ymin)+'):(($'+str(collum)+'*0)'+str(ymin)+') \
             axes x1y1 with candlesticks notitle linecolor rgb "'+items[2]+'"\n\
            '
        if ymin<0 and ymax>0:
            #print (rows_n , collum , frames , x )
            plot='plot  "'+str(infile.name)+'" every ::0::'+str(rows_n)+' using 1:((($'+str(collum)+'/'+str(frames)+')*' +str(x)+') ) axes x1y1 with boxes notitle linecolor rgb "'+items[1]+'",\
            "'+str(infile.name)+'" every ::0::'+str(rows_n)+\
            'using 1:(($'+str(collum)+'*0)+'+str(ymax)+'):('+str(ymax)+'-(('+str(ymax)+'-(($'+str(collum)+' >= 0)  ? $'+str(collum)+' : $'+str(collum)+'*0))/'+str(frames)+')*'+str(x)+'):(($'+str(collum)+'*0)+'+str(ymax)+'):(($'+str(collum)+'*0)+'+str(ymax)+')\
            axes x1y1 with candlesticks notitle linecolor rgb "'+items[0]+'",\
            "'+str(infile.name)+'" every ::0::'+str(rows_n)+\
            ' using 1:(($'+str(collum)+'*0)'+str(ymin)+'):('+str(ymin)+'-(('+str(ymin)+'-(($'+str(collum)+' <= 0)  ? $'+str(collum)+' : $'+str(collum)+'*0))/'+str(frames)+')*' +str(x)+'):(($'+str(collum)+'*0)'+str(ymin)+'):(($'+str(collum)+'*0)'+str(ymin)+') \
            axes x1y1 with candlesticks notitle linecolor rgb "'+items[2]+'"\n\
            '
 #set style fill solid noborder\n\\n\
        cont='\
        set boxwidth 1 relative\n\
        set terminal png font "/usr/openwin/lib/X11/fonts/TrueType/Arial-Bold.ttf" 8\n\
        set output "'+rnd+\
        '"\n\
        set timefmt "'+ str(args.timeformat) +'"\n\
        set xdata time\n\
        set format x"%H:%M"\n\
         \n\
        set xlabel "Cas"\n\
        set ylabel "Hodnota"\n\
         \n\
        set title "'+str(args.legend)+'"\n\
         \n\
        set grid\n\
        set style fill  transparent solid 0.50 noborder\n\
        set yrange ['+str(ymin)+':'+str(ymax)+']\n\
        set xrange [ "'+str(xmin)+'":"'+str(xmax)+'" ]\n\
          '+toset+'\
        '+plot+'\
        ' 
        #Saving configuration             
        with open(filename, 'w') as out:
            out.write(cont)

         #Calling gnuplot
        with open(os.devnull, 'w') as shutup:    
          subprocess.call(['gnuplot',filename], stdout=shutup,stderr=shutup)

  #Finally genereta video file in mp4 format 
  if args.verbouse:
    print ("Genretaing video...\n")     
  with open(os.devnull, 'w') as shutup:
    subprocess.call(['ffmpeg','-r',str(args.fps),'-y','-i',dirpath+'/%d.png',args.name+'.mp4'], stdout=shutup,stderr=shutup)
 
  if args.verbouse:
    print ("Finsihed.... \nCleaning up...") 
 # Clean up temp files
  shutil.rmtree(dirpath)
  if args.verbouse:
    print ("Halting")
  sys.exit(0)

if __name__ == "__main__":
  try:
      main()

  except KeyboardInterrupt:
    shutil.rmtree(dirpath) 
    sys.exit(1)
  except configparser.Error:
    print ("Error in parsing config file")
    shutil.rmtree(dirpath) 
    sys.exit(1)
